itests/eth_deploy_test.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/types/ethtypes"
+	"github.com/filecoin-project/lotus/itests/kit"
+	"github.com/filecoin-project/lotus/node/config"
+)
+
+// TestDeployment smoke tests the deployment of a contract via the
+// Ethereum JSON-RPC endpoint, from an EEOA.
+func TestDeployment(t *testing.T) {
+	// TODO the contract installation and invocation can be lifted into utility methods
+	// He who writes the second test, shall do that.
+	// kit.QuietMiningLogs()
+
+	// reasonable blocktime so that the tx sits in the mpool for a bit during the test.
+	// although this is non-deterministic...
+	blockTime := 1 * time.Second
+	client, _, ens := kit.EnsembleMinimal(
+		t,
+		kit.MockProofs(),
+		kit.ThroughRPC(),
+		kit.WithCfgOpt(func(cfg *config.FullNode) error {
+			cfg.ActorEvent.EnableRealTimeFilterAPI = true
+			return nil
+		}),
+	)
+	ens.InterconnectAll().BeginMining(blockTime)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	// install contract
+	contractHex, err := os.ReadFile("./contracts/SimpleCoin.bin")
+	require.NoError(t, err)
+
+	contract, err := hex.DecodeString(string(contractHex))
+	require.NoError(t, err)
+
+	// create a new Ethereum account
+	key, ethAddr, deployer := client.EVM().NewAccount()
+
+	// send some funds to the f410 address
+	kit.SendFunds(ctx, t, client, deployer, types.FromFil(10))
+
+	// verify balances.
+	bal := client.EVM().AssertAddressBalanceConsistent(ctx, deployer)
+	require.Equal(t, types.FromFil(10), bal)
+
+	// verify the deployer address is an embryo.
+	client.AssertActorType(ctx, deployer, manifest.EmbryoKey)
+
+	gaslimit, err := client.EthEstimateGas(ctx, ethtypes.EthCall{
+		From: &ethAddr,
+		Data: contract,
+	})
+	require.NoError(t, err)
+
+	maxPriorityFeePerGas, err := client.EthMaxPriorityFeePerGas(ctx)
+	require.NoError(t, err)
+
+	// now deploy a contract from the embryo, and validate it went well
+	tx := ethtypes.EthTxArgs{
+		ChainID:              build.Eip155ChainId,
+		Value:                big.Zero(),
+		Nonce:                0,
+		MaxFeePerGas:         types.NanoFil,
+		MaxPriorityFeePerGas: big.Int(maxPriorityFeePerGas),
+		GasLimit:             int(gaslimit),
+		Input:                contract,
+		V:                    big.Zero(),
+		R:                    big.Zero(),
+		S:                    big.Zero(),
+	}
+
+	client.EVM().SignTransaction(&tx, key.PrivateKey)
+
+	pendingFilter, err := client.EthNewPendingTransactionFilter(ctx)
+	require.NoError(t, err)
+
+	hash := client.EVM().SubmitTransaction(ctx, &tx)
+	fmt.Println(hash)
+
+	mpoolTx, err := client.EthGetTransactionByHash(ctx, &hash)
+	require.NoError(t, err)
+
+	// require that the hashes are identical
+	require.Equal(t, hash, mpoolTx.Hash)
+
+	// these fields should be nil because the tx hasn't landed on chain.
+	require.Nil(t, mpoolTx.BlockNumber)
```

Comment:
This can't land as-is. We can either skip the assertion if the msg has already landed (see [here](https://github.com/filecoin-project/lotus/blob/f04aca767f5362868ea3d680c6a3538b4d1babcd/itests/migration_nv17_test.go#L241-L243) for an example), or pause mining between the embryo creation and this assertion (see [here](https://github.com/filecoin-project/lotus/pull/9933#discussion_r1056871841) for more discussion).